unittests/replay_block_invariants_tests.cpp

```diff
@@ -0,0 +1,252 @@
+      BOOST_REQUIRE_NO_THROW(block_log::smoke_test(blocks_dir, 1));
+
+      // retrieve QC block extension from QC block
+      auto block_exts = qc_block->validate_and_extract_extensions();
+      auto qc_ext_itr = block_exts.find(qc_ext_id);
+      auto& qc_ext    = std::get<quorum_certificate_extension>(qc_ext_itr->second);
+      auto& qc        = qc_ext.qc;
+
+      // remove QC block extension from QC block
+      auto& exts = qc_block->block_extensions;
+      std::pair<uint16_t,vector<char>> target{qc_ext_id, {}};
+      auto itr = std::lower_bound(exts.begin(), exts.end(), target, [](const auto& ext1, const auto& ext2){
+         return ext1.first < ext2.first;
+      });
+      BOOST_REQUIRE(itr != exts.end());
+      qc_block->block_extensions.erase(itr);
+
+      // intentionally corrupt QC's signature.
+      auto g2 = qc.active_policy_sig.sig.jacobian_montgomery_le();
+      g2 = bls12_381::aggregate_signatures(std::array{g2, g2});
+      auto affine = g2.toAffineBytesLE(bls12_381::from_mont::yes);
+      qc.active_policy_sig.sig = blslib::bls_aggregate_signature(blslib::bls_signature(affine));
+
+      // add the corrupted QC block extension back to the block
+      emplace_extension(exts, qc_ext_id, fc::raw::pack(qc_ext));
+
+      // add the corrupted block back to block log
+      block_log new_blog(blocks_dir, config.blog);
+      new_blog.append(qc_block, qc_block->calculate_id());
+   }
+
+   // Corrupts finality_extension in the last block of the blocks log
+   // by setting the claimed block number to a different one.
+   void corrupt_finality_extension_in_block_log(uint32_t new_qc_claim_block_num) {
+      controller::config config      = chain.get_config();
+      auto               blocks_dir  = chain.get_config().blocks_dir;
+      auto               fin_ext_id  = finality_extension::extension_id();
+
+      block_log blog(blocks_dir, chain.get_config().blog);
+
+      // retrieve the last block in block log
+      uint32_t  last_block_num = blog.head()->block_num();
+      auto last_block = std::make_shared<signed_block>(blog.read_block_by_num(last_block_num)->clone());
+      BOOST_TEST(last_block);
+
+      // remove last block from block log
+      BOOST_REQUIRE_NO_THROW(block_log::trim_blocklog_end(blocks_dir, last_block_num-1));
+      BOOST_REQUIRE_NO_THROW(block_log::smoke_test(blocks_dir, 1));
+
+      // retrieve finality extension
+      std::optional<block_header_extension> head_fin_ext = last_block->extract_header_extension(fin_ext_id);
+      BOOST_TEST(!!head_fin_ext);
+
+      // remove finality extension from extensions
+      auto& exts = last_block->header_extensions;
+      std::pair<uint16_t,vector<char>> target{fin_ext_id, {}};
+      auto itr = std::lower_bound(exts.begin(), exts.end(), target, [](const auto& ext1, const auto& ext2){
+         return ext1.first < ext2.first;
+      });
+      BOOST_REQUIRE(itr != exts.end());
+      exts.erase(itr);
```

Comment:
I think instead of this lookup just use [std::erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2)